flink-runtime/src/main/java/org/apache/flink/runtime/util/SignalHandler.java

```diff
@@ -53,6 +63,13 @@ public void handle(Signal signal) {
                     "RECEIVED SIGNAL {}: SIG{}. Shutting down as requested.",
                     signal.getNumber(),
                     signal.getName());
+            try {
+                shutdownSupplier
+                        .get()
+                        .get(SHUTDOWN_TIMEOUT.toMilliseconds(), TimeUnit.MILLISECONDS);
```

Comment:
Shouldn't we not wait on `SIGKILL`? Or differently said: Shouldn't we only wait on `SIGTERM`?

Reply:
I do not think `SIGKILL` could be handled via `sun.misc.SignalHandler`. `SIGKILL` means the force-kill, and the JVM does not have any chance to do the signal handler and shutdown hook.

Currently, the `ClusterEntrypoint` will exit when receiving `SIGTERM`, `SIGHUP`, `SIGINT`. So the `shutdownSupplier` will be executed in such cases.

Reply:
I am not aware of other signals that will terminate the JVM process. On Yarn and Kubernetes, when a container needs to be terminated, it will send a `SIGTERM` first and followed by a `SIGKILL` after timeout.

Reply:
I think you are right that the `SignalHandler` only reacts to "TERM", "HUP", "INT". I think I would still prefer a shut down hook because it seems more designed for cleanup purposes.